storage/node.go

```diff
@@ -398,34 +398,22 @@ func (n NodeID) Suffix(prefixBytes, suffixBits int) *Suffix {
 	return NewSuffix(byte(b), sfxPath)
 }
 
-// Prefix returns a copy of NodeID's prefix.
-// This is the same value that would be returned from Split, but without the
-// overhead of calculating the suffix too.
-func (n NodeID) Prefix(prefixBytes int) []byte {
+// Prefix returns a NodeID's prefix of the specified byte length, as another
+// NodeID. Panics if the byte length is out of bound for this ID.
+func (n NodeID) Prefix(bytes int) NodeID {
+	// TODO(pavelkalinnikov): This check doesn't eliminate panics, so we should
+	// either delete it, or cover all other panics.
 	if n.PrefixLenBits == 0 {
-		return []byte{}
+		return NodeID{}
```

Comment:
Map tests are broken because of this line. If I change this to `NodeID{Path: []byte{}}` so that it corresponds to the previous code, tests pass. It's a puzzling puzzle why.

Reply:
OK I'm probably not going to review this fully today so no rush. I've been starting at Merkle trees all week and they're all starting to look the same.

Reply:
After a couple of hours debugging I concluded that it's a bug in MySQL tree storage implementation [here](https://github.com/google/trillian/blob/17aea688cfaaf8d6d64f69982a091903569fdcc8/storage/mysql/tree_storage.go#L212). To read a subtree, it converts `NodeID.Path` directly to the `SubtreeId` field. So, for an empty slice it can either put `nil` or `[]byte{}` to the request. MySQL probably treats `nil` as a SQL `NULL`, while an empty slice corresponds to a blob of length 0.

Reply:
Good news: That bug is probably what caused the existence of `if n.PrefixLenBits == 0` clause here and in a couple of other places in `NodeID`. Now we understand that it should be removed :)

Reply:
Good find. We might have to leave that there unless we know that other storage implementations won't have the same issue. If things would break in an obvious way if they did then maybe we can remove it?